### PR TITLE
[Chore] Refactor and improve Discover field summaries

### PR DESCRIPTION
**Pull request by joshuarrrr**
* Convert to typescript
* Fix types
* Add tests

### Description
Typescript cleanup is a prerequisite to #950
 
### Issues Resolved
N/A
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
    - [x] `yarn test:jest`
    - [x] `yarn test:jest_integration`
    - [x] `yarn test:ftr`
- [ ] New functionality has been documented.
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by joshuarrrr** [comment]
converting to draft until I have time to debug the functional test (I've already manually verified that the the filter controls in field summaries still work with scripted fields, so it's likely something in the test case or test IDs, not in the feature).

**Comment by ashwin-pc** [comment]
Based on our latests discussions, these tests should move to https://github.com/opensearch-project/opensearch-dashboards-functional-test. Lets move it there and deprecate the failing tests.

**Review by bandinib-amzn — DISMISSED** [review]

File: test/functional/apps/management/_scripted_fields.js [hunk 1]
```diff
@@ -509,10 +509,10 @@ export default function ({ getService, getPageObjects }) {
 
       it('should filter by scripted field value in Discover', async function () {
         await PageObjects.discover.clickFieldListItem(scriptedPainlessFieldName2);
-        await log.debug('filter by "Sep 17, 2015 @ 23:00" in the expanded scripted field list');
+        await log.debug('filter by "Sep 18, 2015 @ 7:52" in the expanded scripted field list');
         await PageObjects.discover.clickFieldListPlusFilter(
           scriptedPainlessFieldName2,
-          '1442531297065'
+          '1442562775953'
```
Comment (ashwin-pc) [inline test/functional/apps/management/_scripted_fields.js:515]: Why has this value changed?

File: src/plugins/discover/public/application/components/sidebar/lib/get_details.ts [hunk 2]
```diff
@@ -38,18 +39,30 @@ export function getDetails(
   columns: string[],
   indexPattern?: IndexPattern
 ) {
+  const defaultDetails = {
```
Comment (ashwin-pc) [inline src/plugins/discover/public/application/components/sidebar/lib/get_details.ts:42]: Nice

**Review by ashwin-pc — DISMISSED** [review]

File: test/functional/apps/management/_scripted_fields.js (hunk 1 above)
Comment (joshuarrrr) [reply test/functional/apps/management/_scripted_fields.js:515]: I meant to call this out, but glad you did. This refactor did introduce one slight change to the behavior. When calling `Object.keys(groups)`, the values are naturally sorted, and if they all have the same document count, that natural sort will persist. In the previous implementation, whatever order the values were encountered in would be maintained in a case of a tie, with no natural sorting. From what I could tell, this seemed like just an implementation quirk, and not intended behavior.

File: src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx [hunk 3]
```diff
@@ -49,16 +60,185 @@ const indexPattern = getStubIndexPattern(
...
+
+    await act(async () => {
+      await nextTick();
+      comp.update();
+    });
+    expect(findTestSubject(comp, 'fieldVisualizeLink').length).toBe(0);
+    expect(findTestSubject(comp, 'fieldVisualize-bytes').length).toBe(0);
+  });
+
+  it('should handle promise rejection from getVisualizeHref', async function () {
```
Comment (abbyhu2000) [inline src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx:216]: nice tests!

File: src/plugins/discover/public/application/components/sidebar/lib/get_details.ts [hunk 4]
```diff
@@ -38,18 +39,30 @@ export function getDetails(
...
+    error: '',
+    exists: 0,
+    total: 0,
+    buckets: [],
+    columns,
```
Comment (abbyhu2000) [inline src/plugins/discover/public/application/components/sidebar/lib/get_details.ts:47]: Should we include columns as part of the defaultDetails?

File: src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx [hunk 5]
```diff
@@ -49,16 +60,185 @@ const indexPattern = getStubIndexPattern(
...
+      'bytes'
+    );
+    const comp = mountComponent(visualizableField);
+    expect(findTestSubject(comp, 'fieldVisualizeContainer').length).toBe(1);
+    expect(findTestSubject(comp, 'fieldVisualizeError').length).toBe(0);
+    expect(findTestSubject(comp, 'fieldVisualizeBucketContainer').length).toBe(0);
+    expect(findTestSubject(comp, 'fieldVisualizeLink').length).toBe(0);
+    expect(findTestSubject(comp, 'fieldVisualize-bytes').length).toBe(0);
+
+    await act(async () => {
```
Comment (abbyhu2000) [inline src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx:148]: Just curious what does this act() function do?

File: src/plugins/discover/public/application/components/sidebar/lib/field_calculator.test.ts [hunk 6]
```diff
@@ -187,48 +173,96 @@ describe('fieldCalculator', function () {
...
+      expect(buckets.length).toBe(3);
+      expect(_.map(buckets, 'value')).toEqual(['html', 'gif', 'php']);
       expect(extensions.error).toBe(undefined);
     });
 
     it('fails to analyze geo and attachment types', function () {
-      params.field = indexPattern.fields.getByName('point');
-      expect(fieldCalculator.getFieldValueCounts(params).error).not.toBe(undefined);
+      params.field = indexPattern.fields.getByName('point') as IndexPatternField;
+      expect(getFieldValueCounts(params).error).not.toBe(undefined);
```
Comment (abbyhu2000) [inline src/plugins/discover/public/application/components/sidebar/lib/field_calculator.test.ts:234]: i think toBe(undefined) here can also be replaced by toBeUndefined()? i am not sure which syntax we prefer tho.

**Review by abbyhu2000 — APPROVED** [review]

File: src/plugins/discover/public/application/components/sidebar/lib/get_details.ts (hunk 4 above)
Comment (joshuarrrr) [reply src/plugins/discover/public/application/components/sidebar/lib/get_details.ts:47]: The reason this reads weird is that `columns` is just a value we pass and return without using... which is a bit weird. I can refactor to remove it from this function's input and output and just pass it as a standalone property to `DiscoverFieldDetails` which would be the more sensible thing to do.

File: src/plugins/discover/public/application/components/sidebar/lib/field_calculator.test.ts (hunk 6 above)
Comment (joshuarrrr) [reply src/plugins/discover/public/application/components/sidebar/lib/field_calculator.test.ts:234]: Thanks, nice catch!

File: src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx [hunk 7]
```diff
@@ -49,16 +60,185 @@ const indexPattern = getStubIndexPattern(
...
+        readFromDocValues: true,
+      },
+      'bytes'
+    );
+    const errText = 'Some error';
+    const comp = mountComponent(visualizableField, {
+      details: { ...defaultProps.details, error: errText },
+    });
+    expect(findTestSubject(comp, 'fieldVisualizeContainer').length).toBe(1);
+    expect(findTestSubject(comp, 'fieldVisualizeBucket').length).toBe(0);
```
Comment (joshuarrrr) [inline src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx:179]: This should check the length of the BucketContainer instead.

File: src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx (hunk 5 above)
Comment (joshuarrrr) [reply src/plugins/discover/public/application/components/sidebar/discover_field_details.test.tsx:148]: https://reactjs.org/docs/test-utils.html#act

In this case I introduced it to help distinguish between the component state before and after an async hook runs (see L109-113). There are probably other/better ways to test a case like this, but it was an existing pattern in the codebase. Probably something we want to document better for other similar tests.

**Review by AMoo-Miki — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Review by bandinib-amzn — APPROVED** [review]

**Review by ashwin-pc — APPROVED** [review]

**Pull request merged**